src/Compilers/Core/Portable/Syntax/SyntaxTriviaList.cs

```diff
@@ -236,33 +237,71 @@ public SyntaxTriviaList Insert(int index, SyntaxTrivia trivia)
             return InsertRange(index, new[] { trivia });
         }
 
+        private static readonly ObjectPool<SyntaxTriviaListBuilder> s_builderPool =
+            new ObjectPool<SyntaxTriviaListBuilder>(() => SyntaxTriviaListBuilder.Create());
+
+        private static SyntaxTriviaListBuilder GetBuilder()
+            => s_builderPool.Allocate();
+
+        private static void ClearAndFreeBuilder(SyntaxTriviaListBuilder builder)
+        {
+            // It's possible someone might create a list with a huge amount of trivia
+            // in it.  We don't want to hold onto such items forever.  So only cache
+            // reasonably sized lists.
+            const int MaxBuilderCount = 16;
+            if (builder.Count <= MaxBuilderCount)
+            {
+                builder.Clear();
+                s_builderPool.Free(builder);
+            }
+        }
+
         /// <summary>
         /// Creates a new <see cref="SyntaxTriviaList"/> with the specified trivia inserted at the index.
         /// </summary>
         /// <param name="index">The index in the list to insert the trivia at.</param>
         /// <param name="trivia">The trivia to insert.</param>
         public SyntaxTriviaList InsertRange(int index, IEnumerable<SyntaxTrivia> trivia)
         {
-            if (index < 0 || index > this.Count)
+            var thisCount = this.Count;
+            if (index < 0 || index > thisCount)
             {
                 throw new ArgumentOutOfRangeException(nameof(index));
             }
 
-            var items = trivia.ToList();
-            if (items.Count == 0)
+            if (trivia == null)
             {
-                return this;
+                throw new ArgumentNullException(nameof(trivia));
             }
 
-            var list = this.ToList();
-            list.InsertRange(index, items);
-
-            if (list.Count == 0)
+            // Just return ourselves if we're not being asked to add anything.
+            var triviaList = trivia as ICollection;
+            if (triviaList.Count == 0)
             {
                 return this;
             }
 
-            return new SyntaxTriviaList(default(SyntaxToken), list[0].UnderlyingNode.CreateList(list.Select(n => n.UnderlyingNode)), 0, 0);
+            var builder = GetBuilder();
+            try
+            {
+                for (int i = 0; i < index; i++)
+                {
+                    builder.Add(this[i]);
+                }
+
+                builder.AddRange(trivia);
+
+                for (int i = index; i < thisCount; i++)
+                {
+                    builder.Add(this[i]);
+                }
+
+                return builder.Count == thisCount ? this : builder.ToList();
+            }
+            finally
+            {
+                ClearAndFreeBuilder(builder);
```

Comment:
It still feels like this will be clunky.  In one codepath we don't want to create a new list, but we want to free the builder.  In the other we do want to return a new list, and we do want to free the builder.